### PR TITLE
rgw: remove redundant signature compare in LocalEngine::authenticate

### PR DESCRIPTION
**Pull request by qrGitHub**

### CODE REVIEW

**Comment by qrGitHub** [comment]
Hi @cbodley, would you help to review this? Thanks.

**Comment by qrGitHub** [comment]
Hi @joscollin, would you help to review this? Thanks.

**Comment by yuriw** [comment]
[wip-yuri3-testing-2018-01-02-1716](https://trello.com/c/AqGwkXDb/217-wip-yuri3-testing-2018-01-02-1716)

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -4177,15 +4177,16 @@ rgw::auth::s3::LocalEngine::authenticate(
...
   ldout(cct, 15) << "string_to_sign="
                  << rgw::crypt_sanitize::log_content{string_to_sign}
                  << dendl;
   ldout(cct, 15) << "server signature=" << server_signature << dendl;
   ldout(cct, 15) << "client signature=" << signature << dendl;
-  ldout(cct, 15) << "compare=" << signature.compare(server_signature) << dendl;
+  ldout(cct, 15) << "compare=" << compare << dendl;
 
-  if (static_cast<boost::string_view>(server_signature) != signature) {
+  if (compare != 0) {
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc]: @rzarzynski can you take a look at this one please? was there any material difference between `compare()` and `operator!=()`?

Comment (rzarzynski) [reply src/rgw/rgw_rest_s3.cc]: `VersionAbstractor::server_signature_t` is an alias over our `basic_sstring` template, so it has the overload for conversion to `boost::string_view`. In turn, the `operator!=` of `boost::string_view` internally uses `compare()` enriched with the optimization/early exit for different string sizes.
 
Also, `a != b` appears more readable to me than the `compare()`-based  check.

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -4177,15 +4177,16 @@ rgw::auth::s3::LocalEngine::authenticate(
...
   const VersionAbstractor::server_signature_t server_signature = \
     signature_factory(cct, k.key, string_to_sign);
+  auto compare = signature.compare(server_signature);
 
   ldout(cct, 15) << "string_to_sign="
                  << rgw::crypt_sanitize::log_content{string_to_sign}
                  << dendl;
   ldout(cct, 15) << "server signature=" << server_signature << dendl;
   ldout(cct, 15) << "client signature=" << signature << dendl;
-  ldout(cct, 15) << "compare=" << signature.compare(server_signature) << dendl;
```
Comment (rzarzynski) [inline src/rgw/rgw_rest_s3.cc]: Just a notice: the `compare()` here isn't executed on the main path.

**Comment by qrGitHub** [comment]
Hi @cbodley, can this be merged? Thanks.

**Pull request merged**